controller/tap/apiserver.go

```diff
@@ -0,0 +1,342 @@
+	apiResList, err := apiResourceList()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	clientCAPem, allowedNames, usernameHeader, groupHeader, err := apiServerAuth(k8sAPI)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	server := &apiServer{
+		router:         &httprouter.Router{},
+		k8sAPI:         k8sAPI,
+		apiResList:     apiResList,
+		allowedNames:   allowedNames,
+		usernameHeader: usernameHeader,
+		groupHeader:    groupHeader,
+		tapClient:      client,
+		log: logrus.WithFields(logrus.Fields{
+			"component": "apiserver",
+			"addr":      addr,
+		}),
+	}
+
+	server.router.GET("/apis/"+gvk.GroupVersion().String(), server.handleAPIResourceList)
+
+	for _, res := range resources {
+		route := ""
+		if !res.namespaced {
+			route = fmt.Sprintf("/apis/%s/watch/%s/:namespace/tap", gvk.GroupVersion().String(), res.name)
+		} else {
+			route = fmt.Sprintf("/apis/%s/watch/namespaces/:namespace/%s/:name/tap", gvk.GroupVersion().String(), res.name)
+		}
+		server.router.POST(route, server.handleTap)
+	}
+
+	clientCertPool := x509.NewCertPool()
+	clientCertPool.AppendCertsFromPEM([]byte(clientCAPem))
+
+	wrappedServer := prometheus.WithTelemetry(server)
+
+	s := &http.Server{
+		Addr:    addr,
+		Handler: wrappedServer,
+		TLSConfig: &tls.Config{
+			Certificates: []tls.Certificate{cert},
+			ClientAuth:   tls.VerifyClientCertIfGiven,
+			ClientCAs:    clientCertPool,
+		},
+	}
+
+	lis, err := net.Listen("tcp", addr)
+	if err != nil {
+		server.log.Fatalf("net.Listen failed with: %s", err)
+	}
+
+	return s, lis, nil
+}
+
+// ServeHTTP handles all routes for the APIServer.
+func (a *apiServer) ServeHTTP(w http.ResponseWriter, req *http.Request) {
+	a.log.Debugf("ServeHTTP(): %+v", req)
+
+	// if `requestheader-allowed-names` was empty, allow any CN
+	if len(a.allowedNames) > 0 {
+		validCN := ""
+		clientNames := []string{}
+		for _, cn := range a.allowedNames {
+			for _, clientCert := range req.TLS.PeerCertificates {
+				clientNames = append(clientNames, clientCert.Subject.CommonName)
+				if cn == clientCert.Subject.CommonName {
+					validCN = clientCert.Subject.CommonName
+					break
+				}
+			}
+			if validCN != "" {
+				break
+			}
+		}
+		if validCN == "" {
+			err := fmt.Errorf("no valid CN found. allowed names: %s, client names: %s", a.allowedNames, clientNames)
+			a.log.Debug(err)
+			renderJSONError(w, err, http.StatusBadRequest)
+			return
+		}
+	}
+
+	a.router.ServeHTTP(w, req)
+}
+
+// /apis/tap.linkerd.io/v1alpha1
+// this is required for `kubectl api-resources` to work
+func (a *apiServer) handleAPIResourceList(w http.ResponseWriter, req *http.Request, p httprouter.Params) {
+	w.Header().Set("Content-Type", "application/json")
+	w.Write(a.apiResList)
+}
+
+// /apis/tap.linkerd.io/v1alpha1/watch/namespaces/:namespace/tap
+// /apis/tap.linkerd.io/v1alpha1/watch/namespaces/:namespace/:resource/:name/tap
```

Comment:
It's also possible to make tap requests for a resource type without a resource name.  E.g. `bin/linkerd tap --as trudeau@buoyant.io -n linkerd deploy`

This results in a URL like `/apis/tap.linkerd.io/v1alpha1/watch/namespaces/linkerd/deployments//tap` (note the empty name segment).  Interestingly, the aggregator seems to treat the empty segment as if it's not there: `X-Forwarded-Uri:[/apis/tap.linkerd.io/v1alpha1/watch/namespaces/linkerd/deployments/tap]` and authorizes the user's ability to watch all deployments/tap, which is what we want.  

So, long story short, this all works correctly.  It's just a bit weird that we send along these non-standard URLs.  It might be more "correct" to completely omit the empty segment.

Reply:
Yeah, I was aware of the `deployments//tap` and pleasantly surprised it all just works. I'm inclined to leave it in, as `deployments//tap` makes clear to tap all deployments, while `deployments/tap` could mean either tap all deployments, or tap a single deployment called `tap`.

Reply:
Actually... I wonder how the Linkerd CLI is even able to make these requests and have them get through at all...

This is what happens when I try to make the request manually:

```
$ curl -v -X POST http://localhost:8001/apis/tap.linkerd.io/v1alpha1/watch/namespaces/linkerd/deployments//tap
*   Trying ::1...
* TCP_NODELAY set
* Connection failed
* connect to ::1 port 8001 failed: Connection refused
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to localhost (127.0.0.1) port 8001 (#0)
> POST /apis/tap.linkerd.io/v1alpha1/watch/namespaces/linkerd/deployments//tap HTTP/1.1
> Host: localhost:8001
> User-Agent: curl/7.54.0
> Accept: */*
>
< HTTP/1.1 301 Moved Permanently
< Location: /apis/tap.linkerd.io/v1alpha1/watch/namespaces/linkerd/deployments/tap
< Date: Wed, 31 Jul 2019 22:52:12 GMT
< Content-Length: 0
<
* Connection #0 to host localhost left intact
```

Reply:
Also: `WATCH deployments/tap` watches the deployment named tap, which is not a resource supported by `tap.linkerd.io`.  To tap the deplyoment named tap you'd need to do `WATCH deployments/tap/tap`.